packages/@lwc/integration-karma/test/rendering/callback-invocation-order/index.spec.js

```diff
@@ -297,3 +299,90 @@ it('should invoke callbacks on the right order when multiple templates are used
             ]);
         });
 });
+
+if (process.env.NATIVE_SHADOW) {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    it('regression test (#3827) - shadow DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+        const container = createElement('x-shadow-container-multiple-conditionals', {
+            is: ShadowContainerMultipleConditionals,
+        });
+        document.body.appendChild(container);
+
+        const currentLeafName = container.getLeaf().name;
+        expect(window.timingBuffer).toEqual([
+            'shadowContainer:connectedCallback',
+            'shadowSlot:connectedCallback',
+            `leaf:${currentLeafName}:connectedCallback`,
+        ]);
+
+        resetTimingBuffer();
+
+        const slotCmp = container.getSlotCmp();
+        slotCmp.showIf = true;
+
+        return Promise.resolve()
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+
+                slotCmp.showIf = false;
+                slotCmp.showElseIf = true;
+            })
+            .then(() => {
+                expect(window.timingBuffer).toEqual([]);
+                document.body.removeChild(container);
+            });
+    });
+}
+
+it('regression test (#3827) - light DOM should maintain callback invocation order for duplicate slots across conditional branches when rerender triggered on conditional slot component', () => {
+    // #TODO[3827]: These regression tests can be removed once the bug is resolved.
+    const container = createElement('x-light-container-multiple-conditionals', {
+        is: LightContainerMultipleConditionals,
+    });
+    document.body.appendChild(container);
+
+    let currentLeafName = container.getLeaf().name;
+    expect(window.timingBuffer).toEqual([
+        'lightContainer:connectedCallback',
+        'lightSlot:connectedCallback',
+        `leaf:${currentLeafName}:connectedCallback`,
+    ]);
+
+    resetTimingBuffer();
+
+    const slotCmp = container.getSlotCmp();
+    slotCmp.showIf = true;
+
+    return Promise.resolve()
```

Comment:
You can use async/await in these tests (i.e. `await Promise.resolve()`). It makes things way cleaner IMO.

Reply:
Updated in latest commit!